subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -66,16 +74,46 @@ public ClassSetAnalysis(ClassSetAnalysisData classAnalysis, AnnotationProcessing
         this.resourceDependenciesFromAnnotationProcessing = resourceDependenciesFromAnnotationProcessing.build();
     }
 
-    public ClassSetAnalysis withAnnotationProcessingData(AnnotationProcessingData annotationProcessingData) {
-        return new ClassSetAnalysis(classAnalysis, annotationProcessingData);
+    /**
+     * Computes the types affected by the changes since some other class set, including transitively affected classes.
+     */
+    public ClassSetDiff getChangesSince(ClassSetAnalysis other) {
+        DependentsSet directChanges = classAnalysis.getChangedClassesSince(other.classAnalysis);
+        if (directChanges.isDependencyToAll()) {
+            return new ClassSetDiff(directChanges, Collections.emptyMap());
+        }
+        DependentsSet transitiveChanges = other.getTransitiveDependents(directChanges.getAllDependentClasses(), Collections.emptyMap());
+        if (transitiveChanges.isDependencyToAll()) {
+            return new ClassSetDiff(transitiveChanges, Collections.emptyMap());
+        }
+        DependentsSet allChanges = DependentsSet.merge(Arrays.asList(directChanges, transitiveChanges));
+        Map<String, IntSet> changedConstants = getChangedConstants(other, allChanges);
+        return new ClassSetDiff(allChanges, changedConstants);
+    }
+
+    private Map<String, IntSet> getChangedConstants(ClassSetAnalysis other, DependentsSet affectedClasses) {
+        if (affectedClasses.isDependencyToAll()) {
+            return Collections.emptyMap();
+        }
+        Set<String> dependentClasses = affectedClasses.getAllDependentClasses();
+        Map<String, IntSet> result = new HashMap<>(dependentClasses.size());
+        for (String affectedClass : dependentClasses) {
+            IntSet difference = new IntOpenHashSet(other.getConstants(affectedClass));
+            difference.removeAll(getConstants(affectedClass));
+            result.put(affectedClass, difference);
+        }
+        return result;
     }
 
-    public DependentsSet getRelevantDependents(Iterable<String> classes, IntSet constants) {
-        final Set<String> accessibleResultClasses = new LinkedHashSet<>();
-        final Set<String> privateResultClasses = new LinkedHashSet<>();
-        final Set<GeneratedResource> resultResources = new LinkedHashSet<>();
+    /**
+     * An efficient version of {@link #getTransitiveDependents(String, IntSet)} when several classes have changed.
+     */
+    public DependentsSet getTransitiveDependents(Iterable<String> classes, Map<String, IntSet> constants) {
```

Comment:
I'll do that because I think it adds a little clarity (that we expect a finite collection), but in my profiling sessions I never ever saw this code taking any time at all :)